peps/pep-0747.rst

```diff
@@ -0,0 +1,903 @@
+to a different representation at runtime:
+
+-  For example string-based forward references are normalized at runtime
+   to be ``ForwardRef`` instances: [#forward_ref_normalization]_
+
+   ::
+
+      >>> IntTree = list[typing.Union[int, 'IntTree']]
+      >>> IntTree
+      list[typing.Union[int, ForwardRef('IntTree')]]
+
+-  Special forms themselves often have a runtime representation which is
+   different than how they appear in a source file:
+
+   ::
+
+      >>> list[int]
+      list[int]
+      >>> type(list[int])
+      <class 'types.GenericAlias'>
+      >>> types.GenericAlias(list, int) == list[int]
+      True
+
+The runtime representations of ``TypeExpr``\ s are considered implementation
+details that may change over time and therefore static type checkers are
+not required to recognize them:
+
+::
+
+   INT_TREE: TypeExpr = ForwardRef('IntTree')  # ERROR: Runtime-only form
+   LIST_OF_INT: TypeExpr = types.GenericAlias(list, int)  # ERROR: Runtime-only form
+
+However runtime type checkers must still be prepared to recognize and
+manipulate such runtime-only representations.
+
+Runtime type checkers that wish to assign a runtime-only representation
+of a type expression to a ``TypeExpr[]`` variable must use ``cast()`` to
+avoid errors from static type checkers:
+
+::
+
+   INT_TREE = cast(TypeExpr, ForwardRef('IntTree'))  # OK
+   LIST_OF_INT = cast(TypeExpr, types.GenericAlias(list, int))  # OK
+
+
+Subtyping
+---------
+
+Whether a ``TypeExpr`` value can be assigned from one variable to another is
+determined by the following rules for the is-subtype-of and
+is-consistent-with relationships: [#type_consistency]_
+
+``TypeExpr[]`` is covariant in its argument type, just like ``type[]``:
+
+-  ``TypeExpr[T1]`` is a subtype of ``TypeExpr[T2]`` iff ``T1`` is a
+   subtype of ``T2``.
+-  ``type[C1]`` is a subtype of ``TypeExpr[C2]`` iff ``C1`` is a subtype
+   of ``C2``.
+
+A plain ``type`` can be assigned to a plain ``TypeExpr`` but not the
+other way around:
+
+-  ``type[Any]`` is consistent with ``TypeExpr[Any]``. (But not the
+   other way around.)
+
+``TypeExpr[]`` is a kind of object, just like ``type[]``:
+
+-  ``TypeExpr[T]`` for any ``T`` is a subtype of ``object``.
+
+``TypeExpr[T]``, where ``T`` is a type variable, is assumed to have all
+the attributes and methods of ``object`` and is not callable.
+
+
+.. _interactions_with_type_variables:
+
+Interactions with type variables
+--------------------------------
+
+``TypeExpr[]`` can constrain a type variable that is used elsewhere within
+the same function definition:
+
+::
+
+   def as_instance[T](typx: TypeExpr[T]) -> T | None:
+       return typx() if isinstance(typx, type) else None
+
+A ``TypeExpr[]`` value corresponding to a
+:ref:`non-universal TypeExpr[] <non_universal_typeexpr>` may be
+treated as equivalent to ``Any`` when constraining a type variable:
+
+::
+
+   value = as_instance(TypeIs[str])
+   assert_type(value, Any)  # acceptable
```

Comment:
Type checkers should at least infer `Any | None` (which is not equivalent to Any).

Reply:
Ah. I assumed `Any | None` *was* equivalent to `Any`. Will adjust.